ethcore/types/src/transaction/transaction.rs

```diff
@@ -152,59 +212,134 @@ impl HeapSizeOf for Transaction {
-            value: t.value.into(),
-            data: t.data.into(),
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct AccessListTx {
+    pub transaction: Transaction,
+    //optional access list
+    pub access_list: Vec<(H160, Vec<H256>)>,
+}
+
+impl AccessListTx {
+    pub fn new(transaction: Transaction, access_list: Vec<(H160, Vec<H256>)>) -> AccessListTx {
+        AccessListTx {
+            transaction,
+            access_list,
+        }
+    }
+
+    pub fn tx(&self) -> &Transaction {
+        &self.transaction
+    }
+
+    pub fn tx_mut(&mut self) -> &mut Transaction {
+        &mut self.transaction
+    }
+
+    // decode bytes by this payload spec: rlp([3, [nonce, gasPrice, gasLimit, to, value, data, access_list, senderV, senderR, senderS]])
+    pub fn decode(tx: &[u8]) -> Result<UnverifiedTransaction, DecoderError> {
+        let tx_rlp = &Rlp::new(&tx[1..]); //first byte is related to transaction type defined in EIP-2718
+
+        // we need to have 10 items in this list
+        if tx_rlp.item_count()? != 10 {
+            return Err(DecoderError::RlpIncorrectListLen);
+        }
+        // first part of list is same as legacy transaction and we are reusing that part.
+        let transaction = Transaction::decode_data(&tx_rlp)?;
+
+        // access list we get from here
+        let accl_rlp = tx_rlp.at(6)?;
+
+        // access_list pattern: [[{20 bytes}, [{32 bytes}...]]...]
+        let mut accl: Vec<(H160, Vec<H256>)> = Vec::new();
+
+        for i in 0..accl_rlp.item_count()? {
+            let accounts = accl_rlp.at(i)?;
+
+            if accounts.item_count()? != 2 {
+                //TODO check what to do if we have only one item, should we be strict or not
+                return Err(DecoderError::Custom("Unknown access list lenght"));
+            }
+            accl.push((accounts.val_at(0)?, accounts.list_at(1)?));
+        }
+
+        // we get signature part from here
+        let signature = SignatureComponents {
+            v: tx_rlp.val_at(7)?,
+            r: tx_rlp.val_at(8)?,
+            s: tx_rlp.val_at(9)?,
         };
-        match secret {
-            Some(s) => tx.sign(&s, None),
-            None => tx.null_sign(1),
+
+        //and here we create UnverifiedTransaction and calculate its hash
+        Ok(UnverifiedTransaction::new(
+            TypedTransaction::AccessList(AccessListTx {
+                transaction,
+                access_list: accl,
+            }),
+            signature,
+            0.into(),
+        )
+        .compute_hash())
+    }
+
+    // encode by this payload spec: 0x03 | rlp([3, [nonce, gasPrice, gasLimit, to, value, data, access_list, senderV, senderR, senderS]])
+    pub fn encode(&self, signature: Option<&SignatureComponents>) -> Vec<u8> {
+        let mut stream = RlpStream::new();
+        //stream.begin_list(2);
+        //stream.append(&3u8);
+
+        let mut list_size = 7;
+        list_size += if signature.is_some() { 3 } else { 0 };
```

Comment:
```suggestion
        let list_size = if signature.is_some {
            10
        } else {
            7
        }
```

Reply:
This makes more sense now that we include chain_id.